tracing/trace.go

```diff
@@ -0,0 +1,385 @@
+	sw := tw.root
+	for {
+		ses = append(ses, sw.se)
+		if sw.child == nil {
+			return
+		}
+		sw = sw.child
+	}
+}
+
+func (tw *traceWrapper) findGoroutine(in []stackEntry) []stackEntry {
+	twSes := tw.ses()
+	for t, set := range twSes {
+		if in[0].samePkgMethodStr(set) {
+			merged := append(twSes[0:t+1], in...)
+			for m := range merged[t+1:] {
+				merged[t+m].traceID = twSes[0].traceID
+			}
+			return merged
+		}
+	}
+	return nil
+}
+
+// spanWrapper keeps track of the contexts created for the different spans
+// and closes children as new spans arrive.
+type spanWrapper struct {
+	se     stackEntry
+	ctx    context.Context
+	trace  *traceWrapper
+	parent *spanWrapper
+	child  *spanWrapper
+	hcSpan Span
+}
+
+func newSpanWrapper(ctx context.Context, t *traceWrapper, hcs Span, se stackEntry) *spanWrapper {
+	newSpan := &spanWrapper{
+		ctx:    ctx,
+		trace:  t,
+		hcSpan: hcs,
+		se:     se,
+	}
+	se.addFields(hcs)
+	return newSpan
+}
+
+func newRootSpanWrapper(t *traceWrapper, stack []stackEntry) (root,
+	child *spanWrapper) {
+	root = newSpanWrapper(t.ctx, t, t.hcTrace.GetRootSpan(), stack[0])
+	if len(stack) > 1 {
+		child = root.stackToSpan(stack)
+	} else {
+		child = root
+	}
+	return
+}
+
+func (sw *spanWrapper) add(ns string, val interface{}) {
+	structToFields(sw.hcSpan.AddField, ns, val)
+}
+
+func (sw *spanWrapper) log(lvl int, msg string) {
+	if sw.child != nil {
+		sw.child.done()
+	}
+	_, sLog := sw.hcSpan.CreateChild(sw.ctx)
+	msgStruct := struct {
+		Lvl int
+		Msg string
+	}{lvl, msg}
+	// This value is unfortunately not exported from onet/log/lvl.go
+	if lvl > -17 {
+		structToFields(sLog.AddField, "log", msgStruct)
+	} else {
+		structToFields(sLog.AddField, "error", msgStruct)
+	}
+	sw.se.addFields(sLog)
+	sLog.Send()
+}
+
+func (sw *spanWrapper) done() {
+	if sw.child != nil {
+		sw.child.done()
+	}
+	sw.parent.child = nil
+	sw.hcSpan.Send()
+}
+
+func (sw *spanWrapper) updateSE(se stackEntry) {
+	sw.se = se
+	se.addFields(sw.hcSpan)
+}
+
+func (sw *spanWrapper) stackToSpan(stack []stackEntry) *spanWrapper {
+	if len(stack) == 1 {
+		sw.updateSE(stack[0])
+		return sw
+	}
+	if sw.child != nil {
+		if sw.child.se.traceID != stack[1].traceID ||
+			stack[0].line != sw.se.line {
+			sw.child.done()
+		}
+	}
+	if sw.child == nil {
+		sw.child = sw.createChild(stack[1])
+		sw.updateSE(stack[0])
+	}
+	return sw.child.stackToSpan(stack[1:])
+}
+
+func (sw *spanWrapper) createChild(se stackEntry) *spanWrapper {
+	ctx, newHCS := sw.hcSpan.CreateChild(sw.ctx)
+	newChild := newSpanWrapper(ctx, sw.trace, newHCS, se)
+	newChild.parent = sw
+	return newChild
+}
+
+type stackEntry struct {
+	line      int
+	path      string
+	file      string
+	pkgPath   string
+	method    string
+	params    string
+	traceID   string
+	createdBy bool
+}
+
+var stackFileLine = regexp.MustCompile(
+	"\\s*(.*):([0-9]*)( \\+0x[a-f0-9]*$)?\\s*")
+var stackPkgMethod = regexp.MustCompile(
+	"\\s*(created by )?(.*)\\.(.*?\\w)(\\(.*\\))?\\s*$")
```

Comment:
I might be wrong, but this kind of parsing feels brittle... are there any guarantees Go's stacktraces will keep the same format?

Reply:
Yes, unfortunately it was the only place I could find reliable data about which go-routine is running and the stack-trace. All other calls in `debug`, `runtime`, and `runtime/trace` didn't give me the information I needed. And most of those calls also just return a string and not an actual usable structure.

One thing I did not check is to create an actual trace with `trace.Start` and then evaluate the binary stream and parse it. But I'm not even sure this would be real-time or whether it is just written in junks.

When re-writing, the `trace` library has handy methods - but you'd need to add them throughout the code. And I was looking for a minimal invasive solution.